File: Android/app/src/main/java/com/example/smartplannercompose/domain/dashboard/DashboardStatisticsService.kt
package com.example.smartplannercompose.domain.dashboard

import com.example.smartplannercompose.core.Clock
import com.example.smartplannercompose.core.SystemClock
import com.example.smartplannercompose.domain.tasks.Task
import com.example.smartplannercompose.domain.tasks.TaskPriority
import java.util.Calendar
import java.util.TimeZone

interface DashboardStatisticsService {
    fun statistics(tasks: List<Task>, period: DashboardPeriod): DashboardStatistics
    fun upcoming(tasks: List<Task>, limit: Int): List<Task>
    fun tasksByDay(tasks: List<Task>): Map<Long, List<Task>>
    fun tasks(onDayStartMillis: Long, tasks: List<Task>): List<Task>
    fun startOfDay(millis: Long): Long
}

class DefaultDashboardStatisticsService(
    private val clock: Clock = SystemClock,
    private val timeZone: TimeZone = TimeZone.getDefault()
) : DashboardStatisticsService {

    override fun statistics(tasks: List<Task>, period: DashboardPeriod): DashboardStatistics {
        val now = clock.nowMillis()
        val scoped = tasks.filter { isInPeriod(it, period, now) }

        var completed = 0
        var overdue = 0
        var active = 0
        var noDate = 0
        val byPriority = mutableMapOf<TaskPriority, Int>().apply {
            TaskPriority.values().forEach { put(it, 0) }
        }

        for (task in scoped) {
            byPriority[task.priority] = (byPriority[task.priority] ?: 0) + 1
            if (task.isCompleted) { completed++; continue }
            val deadline = task.deadlineMillis
            if (deadline == null) { noDate++; continue }
            if (deadline < now) overdue++ else active++
        }

        return DashboardStatistics(
            total = scoped.size,
            completed = completed,
            active = active,
            overdue = overdue,
            withoutDate = noDate,
            byPriority = byPriority
        )
    }

    override fun upcoming(tasks: List<Task>, limit: Int): List<Task> {
        val now = clock.nowMillis()
        return tasks
            .asSequence()
            .filter { !it.isCompleted }
            .filter { (it.deadlineMillis ?: now) >= now }
            .sortedBy { it.deadlineMillis ?: Long.MAX_VALUE }
            .take(limit)
            .toList()
    }

    override fun tasksByDay(tasks: List<Task>): Map<Long, List<Task>> {
        val result = HashMap<Long, MutableList<Task>>()
        for (task in tasks) {
            val deadline = task.deadlineMillis ?: continue
            val key = startOfDay(deadline)
            result.getOrPut(key) { mutableListOf() }.add(task)
        }
        return result
    }

    override fun tasks(onDayStartMillis: Long, tasks: List<Task>): List<Task> {
        val target = startOfDay(onDayStartMillis)
        return tasks.filter { task ->
            val d = task.deadlineMillis ?: return@filter false
            startOfDay(d) == target
        }
    }

    override fun startOfDay(millis: Long): Long {
        val c = Calendar.getInstance(timeZone).apply { timeInMillis = millis }
        c.set(Calendar.HOUR_OF_DAY, 0)
        c.set(Calendar.MINUTE, 0)
        c.set(Calendar.SECOND, 0)
        c.set(Calendar.MILLISECOND, 0)
        return c.timeInMillis
    }

    private fun isInPeriod(task: Task, period: DashboardPeriod, now: Long): Boolean {
        if (period == DashboardPeriod.ALL_TIME) return true
        val reference = task.deadlineMillis ?: task.createdAtMillis
        val nowCal = calendarAt(now)
        val refCal = calendarAt(reference)
        return when (period) {
            DashboardPeriod.WEEK ->
                nowCal.get(Calendar.YEAR) == refCal.get(Calendar.YEAR) &&
                    nowCal.get(Calendar.WEEK_OF_YEAR) == refCal.get(Calendar.WEEK_OF_YEAR)
            DashboardPeriod.MONTH ->
                nowCal.get(Calendar.YEAR) == refCal.get(Calendar.YEAR) &&
                    nowCal.get(Calendar.MONTH) == refCal.get(Calendar.MONTH)
            DashboardPeriod.ALL_TIME -> true
        }
    }

    private fun calendarAt(millis: Long): Calendar =
        Calendar.getInstance(timeZone).apply { timeInMillis = millis }
}

File: iOS/SmartPlanner/Domain/Dashboard/DashboardStatisticsService.swift
//
//  DashboardStatisticsService.swift
//  SmartPlanner
//

import Foundation

protocol DashboardStatisticsServicing {
    func statistics(for tasks: [Task], period: DashboardPeriod) -> DashboardStatistics
    func upcoming(from tasks: [Task], limit: Int) -> [Task]
    func tasksByDay(_ tasks: [Task]) -> [Date: [Task]]
    func tasks(on day: Date, from tasks: [Task]) -> [Task]
}

struct DashboardStatisticsService: DashboardStatisticsServicing {

    private let calendar: Calendar
    private let now: () -> Date

    init(calendar: Calendar = .current, now: @escaping () -> Date = Date.init) {
        self.calendar = calendar
        self.now = now
    }

    func statistics(for tasks: [Task], period: DashboardPeriod) -> DashboardStatistics {
        let scoped = tasks.filter { isInPeriod($0, period: period) }
        let reference = now()

        var completed = 0
        var overdue = 0
        var activeWithFutureDate = 0
        var noDate = 0
        var byPriority: [TaskPriority: Int] = [.low: 0, .medium: 0, .high: 0]

        for task in scoped {
            byPriority[task.priority, default: 0] += 1

            if task.isCompleted {
                completed += 1
                continue
            }
            guard let deadline = task.deadline else {
                noDate += 1
                continue
            }
            if deadline < reference {
                overdue += 1
            } else {
                activeWithFutureDate += 1
            }
        }

        return DashboardStatistics(
            total: scoped.count,
            completed: completed,
            active: activeWithFutureDate,
            overdue: overdue,
            withoutDate: noDate,
            byPriority: byPriority
        )
    }

    func upcoming(from tasks: [Task], limit: Int) -> [Task] {
        let reference = now()
        return tasks
            .filter { !$0.isCompleted }
            .filter { ($0.deadline ?? reference) >= reference }
            .sorted { ($0.deadline ?? .distantFuture) < ($1.deadline ?? .distantFuture) }
            .prefix(limit)
            .map { $0 }
    }

    func tasksByDay(_ tasks: [Task]) -> [Date: [Task]] {
        var result: [Date: [Task]] = [:]
        for task in tasks {
            guard let deadline = task.deadline else { continue }
            let day = calendar.startOfDay(for: deadline)
            result[day, default: []].append(task)
        }
        return result
    }

    func tasks(on day: Date, from tasks: [Task]) -> [Task] {
        let target = calendar.startOfDay(for: day)
        return tasks.filter { task in
            guard let deadline = task.deadline else { return false }
            return calendar.startOfDay(for: deadline) == target
        }
    }

    private func isInPeriod(_ task: Task, period: DashboardPeriod) -> Bool {
        switch period {
        case .allTime:
            return true
        case .week:
            return isDate(referenceDate(for: task), inSamePeriodAs: now(), granularity: .weekOfYear)
        case .month:
            return isDate(referenceDate(for: task), inSamePeriodAs: now(), granularity: .month)
        }
    }

    private func referenceDate(for task: Task) -> Date {
        task.deadline ?? task.createdAt
    }

    private func isDate(_ lhs: Date, inSamePeriodAs rhs: Date, granularity: Calendar.Component) -> Bool {
        calendar.isDate(lhs, equalTo: rhs, toGranularity: granularity)
    }
}
